README.md
# NAVER vaccine macro [![Hits](https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2Fchangdoc%2Fnaver-vaccine-macro&count_bg=%233CC2AD&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=hits&edge_flat=false)](https://hits.seeyoufarm.com)

NAVER vaccine macro는 네이버 우리 동네 백신 예약 시도를 자동으로 수행해 주는 크롬 확장 프로그램입니다.

- github 이용 정책 위반으로 프로젝트가 제거되었고 복구가 안된다고 하네요.
## 질문은 [GitLab issue](https://gitlab.com/naver-vaccine-macro/chrome-extensions/-/issues)를 이용해 주세요.

함께 이겨냅시다 코로나가 종식되는 그날까지!

- 잔여 백신 관련 질병관리청 설명 페이지 [코로나19 예방접종 잔여 백신 당일예약 기능 Q&A](http://kdca.go.kr/gallery.es?mid=a20503010000&bid=0002&act=view&list_no=145144)

- 쉬운 설치 [크롬 웹 스토어](https://chrome.google.com/webstore/detail/naver-vaccine-macro/alfhbmpnlhcpcjjaacapcdnggegicepl?hl=ko)

### 사용 방법
1. 크롬 확장 프로그램에서 [NAVER Vaccine MACRO] 설치를 진행합니다.
   - 네이버 웨일 브라우저나 최신 버전을 바로 적용하기 위해서는 개발자 모드로 코드를 받으시고 직접 설치하시면 됩니다.
      - (크롬 개발자 모드 관련 설명 블로그) https://trend21c.tistory.com/1030
1. https://v-search.nid.naver.com/home 네이버 우리 동네 백신 알림 페이지에 접속하여 네이버 인증서를 미리 발급받도록 합니다.
1. 백신 접종을 희망하는 병원을 검색하여 알림 신청을 해둡니다.
1. 잔여 백신이 발생되어 알림을 받게 되면 네이버 톡톡에 [네이버 우리 동네 백신 알림]을 통해 접종기관 알림 전달됩니다. 
1. [네이버 우리 동네 백신 알림] 대화방에서 원하는 접종 기관에서 온 메시지에 있는 [지금 신청하기] 링크를 눌러 이동합니다. (https://v-search.nid.naver.com/reservation/standby?orgCd=XXXX&sid=XXXXX 형태)
   - 알림을 받기 전에 원하는 병원 예약 신청 페이지 url 알아내는 방법은 별도로 작성해두었습니다. 
      - [병원 orgCd, sid 추출 및 예약 신청 페이지 url 추출 방법](https://gitlab.com/naver-vaccine-macro/chrome-extensions/-/wikis/%EB%B3%91%EC%9B%90-orgCd,-sid-%EC%B6%94%EC%B6%9C-%EB%B0%8F-%EC%98%88%EC%95%BD-%EC%8B%A0%EC%B2%AD-%ED%8E%98%EC%9D%B4%EC%A7%80-url-%EC%B6%94%EC%B6%9C-%EB%B0%A9%EB%B2%95))
   - 병원 예약 신청 페이지 url 목록 생성 방법도 작성해두었습니다.
      - [예약 신청 페이지 이동 목록 생성 방법](https://gitlab.com/naver-vaccine-macro/chrome-extensions/-/wikis/%EC%98%88%EC%95%BD-%EC%8B%A0%EC%B2%AD-%ED%8E%98%EC%9D%B4%EC%A7%80-%EC%9D%B4%EB%8F%99-%EB%AA%A9%EB%A1%9D-%EC%83%9D%EC%84%B1-%EB%B0%A9%EB%B2%95)
1. https://v-search.nid.naver.com/reservation/info?key=XXXX 형태의 url, 예약 신청 페이지가 정상적으로 노출되는지 확인합니다.
   - '서비스 이용을 위해 본인인증을 진행해주세요'라는 안내 메시지가 보인다면 본인 인증부터 진행해야 합니다. 
1. 예약 신청 페이지 내에 생성된 '자동 예약 시작' 버튼을 누르면 동작이 시작됩니다.
   - 여러 페이지를 띄워놓고 수행한다면 한 곳에서만 '자동 예약 시작' 버튼을 누르고 나머지 페이지에서는 브라우저에서 새로 고침 하면 동작됩니다.
   - 2021-07-20 부터 네이버 측에서 짧은 시간에 반복해서 새로 고침을 하는 경우 1시간 동안 신청 페이지 접근을 차단 시키므로 한 병원에서만 동작시키시거나 새로 고침 주기를 늘려서 사용하셔야 합니다. (예를 들어 2개 병원 신청 페이지, 새로 고침 주기 5000ms)
     - [관련 내용](https://gitlab.com/naver-vaccine-macro/chrome-extensions/-/wikis/%EC%9E%A0%EC%8B%9C-%ED%9B%84-%EB%8B%A4%EC%8B%9C-%EC%8B%9C%EB%8F%84%ED%95%98%EC%84%B8%EC%9A%94.-(%EC%98%88%EC%95%BD-%EC%8B%A0%EC%B2%AD-%ED%8E%98%EC%9D%B4%EC%A7%80-%EC%B0%A8%EB%8B%A8)-%EA%B4%80%EB%A0%A8)
1. 페이지를 계속 새로 고침 하며 잔여 수량이 있을 경우 예약을 시도합니다. (test mode 체크된 상태에서는 실제로 시도하지 않습니다.)
1. 예약이 시도 시 소리와 함께 결과 화면으로 이동합니다.
1. 네이버 맵, 카카오 맵 에서 동작하는 매크로 기능은 가이드 문서에서 확인해주세요.
      - [네이버 맵](https://gitlab.com/naver-vaccine-macro/chrome-extensions/-/wikis/%EB%84%A4%EC%9D%B4%EB%B2%84-%EB%A7%B5%EC%97%90%EC%84%9C-%EB%A7%A4%ED%81%AC%EB%A1%9C-%EB%8F%99%EC%9E%91-%EB%B0%A9%EB%B2%95)
      - [카카오 맵](https://gitlab.com/naver-vaccine-macro/chrome-extensions/-/wikis/%EC%B9%B4%EC%B9%B4%EC%98%A4-%EB%A7%B5%EC%97%90%EC%84%9C-%EB%A7%A4%ED%81%AC%EB%A1%9C-%EB%8F%99%EC%9E%91-%EB%B0%A9%EB%B2%95)

### 주의 사항
1. 예약 시도 시 사운드가 재생되므로 스피커를 켜 두는게 좋습니다.
1. 예약 시도 후에는 자동 예약 프로그램은 정지됩니다.
1. 예약 시도가 반드시 성공한다는 보장이 없습니다.
1. 예약 시도 후에는 다시 시도할 경우 사용 방법 5번째 항목에 있는 [지금 신청하기] 링크를 눌러 이동하는 스텝부터 다시 수행해야 합니다.

### Version History
https://gitlab.com/naver-vaccine-macro/chrome-extensions/-/blob/main/CHANGELOG.md

### 
본 프로그램은 KTX Macro
[KTX Macro github page](https://github.com/youngjin-k/ktx-macro) 
에서 Fork되어 제작되었습니다.

### License
MIT License


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- 네이버 맵 새로고침 html 구조 변경 대응
- 네이버 맵 api response 변경사항 반영
- 네이버 플레이스 잔여백신 링크 변경 https://m.place.naver.com/rest/vaccine

## [1.1.6] - 2021-08-12
- 예약 신청 페이지에서 잔여백신 있지만 대상이 아닐때 동작 멈추는 이슈 수정
- 맵에서 아무거나 선택시 예약 신청 페이지로 이동 안하고 있는 부분 수정

## [1.1.5] - 2021-08-10
- 맵(카카오 잔여 백신)에서 선택 백신 종류로 탐색하도록 로직 추가 (카카오 인증 후 동작 필요)
- 카카오 맵에서 잔여 백신 종류 선택 상황에서 개별 병원 조회 시도 수가 없을때 동작 멈추는 부분 수정

## [1.1.4] - 2021-08-09
- 맵(네이버 플레이스 잔여 백신)에서 선택 백신 종류로 탐색하도록 로직 추가

## [1.1.3] - 2021-08-06
- 본인 인증 요구되는 상황에서 매크로 동작, 버튼 노출 하지 않도록 수정

## [1.1.2]
- 네이버 맵에서 네이버 graphql로 생성한 병원 목록에 존재하지 않는 경우 멈추고 있는 이슈 수정
- 카카오 맵 잔여 백신 조회 api v2 -> v3로 변경

## [1.1.1] - 2021-07-31
### Changed
- 맵에서 잔여 백신 탐지했으나 네이버 graphql로 생성한 병원 목록에 존재하지 않는 경우 멈추지 않고 새로 고침 계속 수행되도록 수정 

## [1.1.0] - 2021-07-27
### Added
- 맵(네이버 플레이스 잔여 백신, 카카오 잔여 백신)에서 새로 고침 수행하는 스크립트 추가
### Changed
- 작동 버튼에 specific class 제거, inline style 적용
- chrome extension manifest V3로 마이그레이션 -> 웨일에서 동작 안 해서 V2로 다시 롤백
- 사운드 성공 시에만 tada, 나머지는 boing 재생
- chrome 스토어에서 tabs permission 사용되는 곳 없다고 리젝 하여 permission 삭제

## [1.0.9] - 2021-07-20
### Changed
- localStorage 사용하지 않고 chrome.storage사용하여 동작 여부 체크
- 작동 버튼 class 랜덤 문자열로 생성
- 기본,최소 인터벌 2000ms

## [1.0.8] - 2021-07-19
### Changed
- 병원 이름 포함된 텔레그램 메시지 전송 예약 시도 타이밍에 추가
- 예약 신청 방식 다시 변경 (예약 신청 버튼 click event 발생시키는 형태 -> url 변경하여 페이지 이동)

## [1.0.7] - 2021-07-19
### Changed
- 매크로 버튼 추가 동작 타이밍 조금 늦춤
- 텔레그램 메시지 전송시 예약 시도 병원 이름 포함 되도록

## [1.0.6] - 2021-07-16
### Added
- 모든 링크 열기 버튼 추가
### Changed
- 새로고침 주기 저장 후 리로드 시 잘 못 표기 되고 있는 부분 수정
- 새로고침 주기 반영시 0~20ms씩 변화 있도록 동작하게 수정
- 예약신청 페이지에서 에러 감지 되었을 때는 사운드 재생 되지 않도록 수정
- orgCd만으로는 예약 신청 url 접근 불가하여 sid 파라미터 추가

## [1.0.5] - 2021-07-13
### Added
- 링크 여러개 한번에 열기 기능 추가 [Issue#11](https://github.com/changdoc/naver-vaccine-macro/issues/11).
- 백신 선택 옵션 모더나 추가

## [1.0.4.1] - 2021-07-11
### Changed
- 에러 페이지 노출시 매크로 동작 멈춤 증상 수정 FINAL

## [1.0.4] - 2021-07-11
### Changed
- 에러 페이지 검사 시점 백신 종류 리스트 확인 후에 없을때만 수행되도록 변경
- 에러 페이지 검출 됐을때도 확인을 위해 텔레그램 메시지 발송

## [1.0.3] - 2021-07-10
### Added
- graphql response 처리하여 예약 신청 페이지 이동 편하게 할 수 있도록 목록 생성 기능 추가

### Changed
- 예약 신청 방식 변경 (url 변경하여 페이지 이동 -> 예약 신청 버튼 click event 발생시키는 형태)
- 예약 신청 시도 됐을때 결과를 알 수 없는 상태 (질병 관리 본부 측 timeout)에 텔레그램 메시지 구분하여 전송 되게 수정

### Removed
- 본문에 에러 항목 있을때 새로고침 처리 하는 부분 제거 (오동작 방지 추후 문제 없으면 다시 동작하도록 변경 예정)

## [1.0.2] - 2021-07-09
### Added
- 자동 예약 시작 안내 메시지에 여러 창 떠있을때 새로고침 안내 메시지 추가
- 개인정보 제공동의 input 있으면 자동 checked

### Changed
- error 페이지 노출로 판단되어도 매크로 멈추지 않도록 수정
- 예약 결과 텔레그램 메시지 못보내고 있던 부분 수정

## [1.0.1] - 2021-07-08
### Added
- 백신 선택 옵션 화이자 추가
- test 시 시도 했을 예약 시도 url 디버깅 용으로 추가

### Changed
- test off, 예약 시도 될 때 테스트 메시지 전송 되지 않도록 수정

## [1.0.0] - 2021-06-23
- Initial Version


map.js
let sButtonText = "";
let sButtonReload = undefined;
let _tick;
const RELOAD_INTERVAL_MILLISECONDS = 200;
let hospitalList;
let hospitalListStrCache;
const _className = "_" + Math.random().toString(36).substring(2, 7);
const _buttonStyle = "width: 175px;height: 38px;line-height: 38px;text-align: center;background: linear-gradient(to bottom, #17e677, #19d1d1);color: #ffffff;font-size: 13px;border: 1px solid #149393;cursor: pointer;float:right;";

let nums = new Array();

const injectButton = () => {
    if (!sButtonReload)
        return;

    sButtonReload.parentNode.insertAdjacentHTML(
        "beforeend",
        `
        <span><button type="button" class="${_className}" style="${_buttonStyle}">${sButtonText}</button></span>
    `
    );

    document
        .querySelector(`.${_className}`)
        .addEventListener("click", onMacroClick);
}

const setEscapeEvent = () => {
    window.addEventListener("keydown", e => {
        if (e.key === "Escape") {
            macroStop();
        }
    });
};

const onMacroClick = () => {
    chrome.storage.local.get(function (data) {
        const isStarted = (data.macro === "on");
        if (isStarted)
            macroStop();
        else
            macroStart();
    });

    // const isStarted = localStorage.getItem("macro") === "on";
    // if (isStarted)
    //     macroStop();
    // else macroStart();
};

const macroStart = () => {
    alert(
        "자동 새로고침을 시작합니다.\n" +
        "예약 시도 시 예약 신청 페이지로 이동합니다.\n" +
        "탐지할 지역의 graphql result을 입력해두어야 예약 신청 페이지로 이동 가능합니다.\n" +
        "자동 새로고침 종료는 '자동 새로고침 정지' 혹은 esc키를 눌러주세요."
    );

    chrome.storage.local.set({macro: "on"}, function () {
        reload();
    });

    let button = document.querySelector(`.${_className}`);
    if (button) {
        button.innerText = "자동 새로고침 정지(ESC)";
    }
    setEscapeEvent();
    // localStorage.setItem("macro", "on");
};

const macroStop = (result) => {
    if (!result)
        alert("자동 새로고침을 종료합니다.");

    chrome.storage.local.set({macro: ""});

    clearTimeout(_tick);

    let button = document.querySelector(`.${_className}`);
    if (button) {
        button.innerText = "자동 새로고침 다시 시작";
    }
};

const _reload = () => {
    let settingInterval = RELOAD_INTERVAL_MILLISECONDS;

    let modify = ((new Date().getTime()) % 20);

    // console.log("setting interval:" + settingInterval + ", modify:" + modify);

    if (settingInterval > modify)
        settingInterval -= modify;

    // console.log("curr interval:" + settingInterval);

    _tick = setTimeout(reload, settingInterval);
};

const reload = () => {
    // recover numbers color
    if (nums.length > 0) {
        for (let i = 0; i < nums.length; i++) {
            const num = nums[i];
            num.style.backgroundColor = "";
        }
        nums.length = 0;
    }

    // console.log(sButtonReload);
    // console.log(sButtonReload.classList);
    // console.log(sButtonReload.classList.length);
    if (sButtonReload)
        sButtonReload.click();
    else {
        // console.log("will reloading ignore current tick.");
    }
};

const afterReload = () => {
    chrome.storage.local.get(function (value) {
        const isStarted = value.macro === "on";

        if (isStarted) {
            macro(value);
        }

        if (!sButtonReload) {
            console.log("can't found reload button, ignored macro.");
            return;
        }
    });
};

function _findWithSelectedName(selected) {
    const selOnMenu = selected;
    if (selected == '아스트라제네카')
        selected = 'AZ';

    let listContainer = document.getElementById('_list_scroll_container');
    let placesForSelect = listContainer.querySelectorAll('.place_blind');
    // console.log(placesForSelect);
    let checkHospitalCount = 0;
    for (let i = 0; i < placesForSelect.length; i++) {
        if (placesForSelect[i].innerHTML == '잔여백신종류') {
            let next = placesForSelect[i].parentNode.nextSibling;
            if (!next) continue;

            while (next) {
                let curr = next;
                next = curr.nextSibling;

                let vaccineName = curr.firstChild.nodeValue;
                if (selected != vaccineName)
                    continue;

                let targetName = curr.parentNode.parentNode.childNodes[0].querySelector('span').innerText;
                let em = curr.querySelector('em');
                let num = parseInt(em.firstChild.nodeValue);
                if (isNaN(num))
                    num = 0;
                // console.log(targetName, num, curr ? curr.innerText : '','/', curr ? curr.innerHTML : '','/', curr ? curr.firstChild : '','/', em,'/', em ? em.firstChild : '');

                let enable = num > 0;
                if (enable) {
                    console.log(targetName, '/', vaccineName, '/', num, enable);
                    let info = getHospital(targetName);
                    if (openReservePage(targetName, info))
                        return;
                }
                else
                    ++checkHospitalCount;
            }
        }
    }
    console.log("선택 백신 종류:", selOnMenu, ",현재 확인 병원 수 :", checkHospitalCount);

    // keep going
    _reload();
}

const macro = (data) => {
    const hospitalListStr = data.hospital_list;
    // console.log("execute macro." + hospitalListStr);
    if (hospitalListStr && hospitalListStrCache != hospitalListStr) {
        hospitalList = JSON.parse(hospitalListStr);
        hospitalListStrCache = hospitalListStr;
        // console.log("parsed list. hospital list string changed.");
    }

    // console.log(hospitalList);
    if (!hospitalList || hospitalList.length == 0) {
        console.log("hospitalList is empty. cannot macro execute.");
        macroStop('empty');
        alert("병원 목록 데이터가 없습니다. 현재 지도에서 새로고침을 수행한 graphql result를 매크로 페이지에 입력하여 병원 목록을 생성한 뒤 동작 시켜주세요.");
        return;
    }

    let selectedName = null;
    if (data.selected_vaccine != null)
        selectedName = data.selected_vaccine;

    if (selectedName != null && selectedName != undefined && selectedName.length > 0) {
        _findWithSelectedName(selectedName);
        return;
    }

    let places = document.querySelectorAll(".place_blind");
    // console.log(places);
    let arr = new Array();

    for (let i = 0; i < places.length; i++) {
        // console.log(places[i].parentNode);
        if (places[i].parentNode.querySelectorAll(".place_blind").length >= 3) {
            arr.push(places[i].parentNode);
        }
    }

    if (arr.length > 0) {
        for (let i = 0; i < arr.length; i++) {
            // console.log(arr[i]);
            let targetName = arr[i].childNodes[0];
            let targetNum = arr[i].childNodes[arr[i].childNodes.length - 1];
            // console.log(targetNum);
            // console.log(targetNum.innerText);
            const name = targetName.innerText;
            const num = parseInt(targetNum.innerText);

            let info = getHospital(name);

            // console.log(name + ", " + num);
            // console.log(info);

            let enable = num > 0;
            // targetName.style.backgroundColor = enable ? "#3ef03e" : "#f03e3e";
            targetNum.style.backgroundColor = enable ? "#3ef03e" : "#f03e3e";

            nums.push(targetNum);

            if (enable) {
                if (openReservePage(name, info))
                    return;
            }
        }
    }

    // keep going
    _reload();
};

const RESERVE_URL = 'https://v-search.nid.naver.com/reservation?orgCd=';
const RESERVER_URL_ADD_PARM = '&sid=';

const openReservePage = (name, obj) => {
    if (obj && obj.vaccineQuantity && obj.vaccineQuantity.vaccineOrganizationCode) {
        const orgCd = obj.vaccineQuantity.vaccineOrganizationCode;
        const sid = obj.id;
        const url = RESERVE_URL + orgCd + RESERVER_URL_ADD_PARM + sid;
        console.log("try open url : " + url);
        document.location.href = url;
        return true;
    } else {
        console.log('cannot found open url for reserve. ignored : ' + name);
        return false;
    }
}

const getHospital = (name) => {
    if (!hospitalList)
        return undefined;
    for (let i = 0; i < hospitalList.length; ++i) {
        if (hospitalList[i].name == name) {
            return hospitalList[i];
        }
    }
    return undefined;
};

let reloading = false;
const observer = new MutationObserver(function (mutations) {
    mutations.forEach(function (mutation) {
        if (mutation.type == "attributes") {
            let prevReloading = reloading;
            if (sButtonReload)
                reloading = sButtonReload.classList.length == 2;

            if (prevReloading && !reloading) {
                // console.log("reload done. try after reload.");
                // setTimeout(afterReload, 0);
                afterReload();
            }
        }
    });
});

const _prepare = () => {
    chrome.storage.local.get(function (value) {

        observer.observe(sButtonReload, {
            attributes: true //configure it to listen to attribute changes
        });

        const isStarted = value.macro === "on";

        if (isStarted) {
            macro(value);
            setEscapeEvent();
        }

        sButtonText = isStarted ? "자동 새로고침 정지(ESC)" : "자동 새로고침 시작";
        setTimeout(injectButton, 100);
    });
};

const _assureReloadButton = () => {
    let places = document.querySelectorAll("span");
    // console.log(places);
    for (let i = 0; i < places.length; i++) {
        if (!sButtonReload && places[i].innerHTML == "새로고침") {
            sButtonReload = places[i].parentNode;
            break;
        }
    }
};

let tryCount = 0;
const _prepareIf = () => {
    _assureReloadButton();

    if (!sButtonReload) {
        if (++tryCount > 5)
            console.log(`can't found reload button ${tryCount}times, try again after 500ms.`);
        setTimeout(_prepareIf, 500);
        return;
    }

    _prepare();
};

(() => {
    // console.log("map.js loaded.");
    _prepareIf();
})();

map_kakao.js
let sButtonText = "";
let sButtonReload = undefined;
let _tick;
const RELOAD_INTERVAL_MILLISECONDS = 500;
let hospitalList;
let hospitalListStrCache;
const _className = "_" + Math.random().toString(36).substring(2, 7);
const _buttonStyle = "width: 175px;height: 38px;line-height: 38px;text-align: center;background: linear-gradient(to bottom, #17e677, #19d1d1);color: #ffffff;font-size: 13px;border: 1px solid #149393;cursor: pointer;";

let nums = new Array();

const injectButton = () => {
    if (!sButtonReload)
        return;

    sButtonReload.parentNode.insertAdjacentHTML(
        "beforeend",
        `
        <span><button type="button" class="${_className}" style="${_buttonStyle}">${sButtonText}</button></span>
    `
    );

    document
        .querySelector(`.${_className}`)
        .addEventListener("click", onMacroClick);
}

const setEscapeEvent = () => {
    window.addEventListener("keydown", e => {
        if (e.key === "Escape") {
            macroStop();
        }
    });
};

const onMacroClick = () => {
    chrome.storage.local.get(function (data) {
        const isStarted = (data.macro === "on");
        if (isStarted)
            macroStop();
        else
            macroStart();
    });

    // const isStarted = localStorage.getItem("macro") === "on";
    // if (isStarted)
    //     macroStop();
    // else macroStart();
};

const macroStart = () => {
    alert(
        "자동 새로고침을 시작합니다.\n" +
        "예약 시도 시 예약 신청 페이지로 이동합니다.\n" +
        "탐지할 지역의 graphql result을 입력해두어야 예약 신청 페이지로 이동 가능합니다.\n" +
        "자동 새로고침 종료는 '자동 새로고침 정지' 혹은 esc키를 눌러주세요."
    );

    chrome.storage.local.set({macro: "on"}, function () {
        reload();
    });

    let button = document.querySelector(`.${_className}`);
    if (button) {
        button.innerText = "자동 새로고침 정지(ESC)";
    }
    setEscapeEvent();
    // localStorage.setItem("macro", "on");
};

const macroStop = (result) => {
    if (!result)
        alert("자동 새로고침을 종료합니다.");

    chrome.storage.local.set({macro: ""});

    clearTimeout(_tick);

    let button = document.querySelector(`.${_className}`);
    if (button) {
        button.innerText = "자동 새로고침 다시 시작";
    }
};

const _reload = () => {
    let settingInterval = RELOAD_INTERVAL_MILLISECONDS;

    let modify = ((new Date().getTime()) % 20);

    // console.log("setting interval:" + settingInterval + ", modify:" + modify);

    if (settingInterval > modify)
        settingInterval -= modify;

    // console.log("curr interval:" + settingInterval);

    _tick = setTimeout(reload, settingInterval);
};

const reload = () => {

    // recover numbers color
    if (nums.length > 0) {
        for (let i = 0; i < nums.length; i++) {
            const num = nums[i];
            num.style.backgroundColor = "";
        }
        nums.length = 0;
    }

    // console.log(sButtonReload);
    // console.log(sButtonReload.classList);
    // console.log(sButtonReload.classList.length);
    // if (sButtonReload)
    //     sButtonReload.click();
    // else {
    //     // console.log("will reloading ignore current tick.");
    // }

    sButtonReload.classList.add("refresh_animation_step1");
    setTimeout(function () {
        sButtonReload.classList.add("refresh_animation_step2");
    }, 200);

    chrome.storage.local.get(function (value) {
        let body = value.kakao_body;
        if (!body) {
            alert("left_count_by_coords api호출의 request body 데이터가 필요합니다.");
            macroStop('data_error');
            return;
        }

        fetch("https://vaccine-map.kakao.com/api/v3/vaccine/left_count_by_coords", {
            "headers": {
                "accept": "application/json, text/plain, */*",
                "accept-language": "ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7",
                "content-type": "application/json;charset=UTF-8",
                "sec-ch-ua": "\" Not;A Brand\";v=\"99\", \"Google Chrome\";v=\"91\", \"Chromium\";v=\"91\"",
                "sec-ch-ua-mobile": "?0",
                "sec-fetch-dest": "empty",
                "sec-fetch-mode": "cors",
                "sec-fetch-site": "same-origin"
            },
            "referrer": "https://vaccine-map.kakao.com/map2",
            "referrerPolicy": "strict-origin-when-cross-origin",
            "body": body,
            "method": "POST",
            "mode": "cors",
            "credentials": "include"
        }).then(function (response) {
            sButtonReload.classList.remove("refresh_animation_step1");
            sButtonReload.classList.remove("refresh_animation_step2");
            afterReload(response);
        });
    });
};

const afterReload = (response) => {
    chrome.storage.local.get(function (value) {
        const isStarted = value.macro === "on";

        if (isStarted) {
            macro(value, response);
        }

        if (!sButtonReload) {
            console.log("can't found reload button, ignored macro.");
            return;
        }
    });
};

function _findWithSelectedName(selected, data) {
    const length = data === undefined || data.organizations === undefined ? 0 : data.organizations.length;
    const selOnMenu = selected;

    let loadedCount = 0;
    let loadCount = 0;
    for (let i = 0; i < length; i++) {
        const curr = data.organizations[i];
        const enable = curr.leftCounts > 0;

        if (!enable) {
            continue;
        }

        console.log("curr.leftCounts > 0, " + curr.orgName);

        let info = getHospitalByCode(curr.orgCode);
        if (info == null) {
            console.log("found by orgCode is null : " + curr);
            info = getHospital(curr.orgName);
        }

        ++loadCount;
        const countUrl = '/api/v3/org/org_code/' + curr.orgCode;

        fetch(countUrl, {
            "headers": {
                "accept": "application/json, text/plain, */*",
                // "accept-language": "ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7",
                "content-type": "application/json;charset=UTF-8",
                // "sec-ch-ua": "\" Not;A Brand\";v=\"99\", \"Google Chrome\";v=\"91\", \"Chromium\";v=\"91\"",
                // "sec-ch-ua-mobile": "?0",
                // "sec-fetch-dest": "empty",
                // "sec-fetch-mode": "cors",
                // "sec-fetch-site": "same-origin"
            },
            // "referrer": "https://vaccine-map.kakao.com/map2",
            // "referrerPolicy": "strict-origin-when-cross-origin",
            "method": "GET",
            // "mode": "cors",
            // "credentials": "include"
        }).then(function (response) {
            if (response && response.status === 200) {
                response.json().then(function (data) {
                    if (data.leftCount > 0) {
                        for (let i = 0; i < data.lefts.length; ++i) {
                            if (data.lefts[i].vaccineName == selOnMenu
                                && data.lefts[i].leftCount > 0) {
                                if (openReservePage(curr.orgName, info))
                                    return;
                            }
                        }
                    }

                    if (++loadedCount == loadCount)
                    {
                        console.log("선택 백신 종류:", selOnMenu, ",현재 확인 병원 수 :", loadedCount);
                        if (!movingToReserve)
                        {
                            // keep going
                            _reload();
                        }
                    }
                });
            }
            else
            {
                if (++loadedCount == loadCount)
                {
                    console.log("선택 백신 종류:", selOnMenu, ",현재 확인 병원 수 :", loadedCount);
                    if (!movingToReserve)
                    {
                        // keep going
                        _reload();
                    }
                }
            }
        });
    }

    if (loadCount == 0)
        _reload();
}

const macro = (value, response) => {
    const hospitalListStr = value.hospital_list;
    // console.log("execute macro." + hospitalListStr);
    if (hospitalListStr && hospitalListStrCache != hospitalListStr) {
        hospitalList = JSON.parse(hospitalListStr);
        hospitalListStrCache = hospitalListStr;
        // console.log("parsed list. hospital list string changed.");
    }

    // console.log(hospitalList);
    if (!hospitalList || hospitalList.length == 0) {
        console.log("hospitalList is empty. cannot macro execute.");
        macroStop('empty');
        alert("병원 목록 데이터가 없습니다. 현재 지도에서 새로고침을 수행한 graphql result를 매크로 페이지에 입력하여 병원 목록을 생성한 뒤 동작 시켜주세요.");
        return;
    }

    let selectedName = null;
    if (value.selected_vaccine != null)
        selectedName = value.selected_vaccine;

    if (response && response.status === 200) {
        response.json().then(function (data) {
            // console.log(data);
            if (selectedName != null && selectedName != undefined && selectedName.length > 0) {
                _findWithSelectedName(selectedName, data);
                return;
            }

            const length = data.organizations.length;
            for (let i = 0; i < length; i++) {
                const curr = data.organizations[i];
                const enable = curr.leftCounts > 0;
                // console.log(curr.orgCode, curr.leftCounts, curr.orgName);

                if (enable) {
                    console.log("curr.leftCounts > 0, " + curr.orgName);

                    let info = getHospitalByCode(curr.orgCode);
                    if (info == null) {
                        console.log("found by orgCode is null : " + curr);
                        info = getHospital(curr.orgName);
                    }

                    if (openReservePage(curr.orgName, info))
                        return;
                }
            }

            // keep going
            _reload();
        });
        return;
    }

    // keep going
    _reload();
};

const RESERVE_URL = 'https://v-search.nid.naver.com/reservation?orgCd=';
const RESERVER_URL_ADD_PARM = '&sid=';
let movingToReserve = false;
const openReservePage = (name, obj) => {
    if (movingToReserve) {
        console.log("already moving..!");
        return false;
    }
    if (obj && obj.vaccineQuantity && obj.vaccineQuantity.vaccineOrganizationCode) {
        movingToReserve = true;
        const orgCd = obj.vaccineQuantity.vaccineOrganizationCode;
        const sid = obj.id;
        const url = RESERVE_URL + orgCd + RESERVER_URL_ADD_PARM + sid;
        console.log("try open url : " + url);
        document.location.href = url;
        return true;
    } else {
        console.log('cannot found open url for reserve. ignored : ' + name);
        return false;
    }
}

const getHospital = (name) => {
    if (!hospitalList)
        return undefined;
    for (let i = 0; i < hospitalList.length; ++i) {
        if (hospitalList[i].name == name) {
            return hospitalList[i];
        }
    }
    return undefined;
};

const getHospitalByCode = (code) => {
    if (!hospitalList)
        return undefined;
    for (let i = 0; i < hospitalList.length; ++i) {
        const org = hospitalList[i];
        if (org && org.vaccineQuantity && org.vaccineQuantity.vaccineOrganizationCode == code) {
            return hospitalList[i];
        }
    }
    return undefined;
};

// let reloading = false;
// const observer = new MutationObserver(function (mutations) {
//     mutations.forEach(function (mutation) {
//         if (mutation.type == "attributes") {
//             let prevReloading = reloading;
//             if (sButtonReload)
//                 reloading = sButtonReload.classList.length >= 2;
//
//             if (prevReloading && !reloading) {
//                 // console.log("reload done. try after reload.");
//                 // setTimeout(afterReload, 0);
//                 afterReload();
//             }
//         }
//     });
// });

const _prepare = () => {
    chrome.storage.local.get(function (value) {

        // observer.observe(sButtonReload, {
        //     attributes: true //configure it to listen to attribute changes
        // });

        const isStarted = value.macro === "on";

        if (isStarted) {
            macro(value);
            setEscapeEvent();
        }

        sButtonText = isStarted ? "자동 새로고침 정지(ESC)" : "자동 새로고침 시작";
        setTimeout(injectButton, 100);
    });
};

const _assureReloadButton = () => {
    if (!sButtonReload)
        sButtonReload = document.querySelector(".btn_refresh");
};

let tryCount = 0;
const _prepareIf = () => {
    _assureReloadButton();

    if (!sButtonReload) {
        if (++tryCount > 5)
            console.log(`can't found reload button ${tryCount}times, try again after 500ms.`);
        setTimeout(_prepareIf, 500);
        return;
    }

    _prepare();
};

(() => {
    // console.log("map.js loaded.");
    _prepareIf();
})();